File: java/say/src/main/java/Say.java
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public class Say {
  private static final String[] scales = new String[] {null, "thousand", "million", "billion"};
  private static final String[] toTwenty =
      new String[] {
        null,
        "one",
        "two",
        "three",
        "four",
        "five",
        "six",
        "seven",
        "eight",
        "nine",
        "ten",
        "eleven",
        "twelve",
        "thirteen",
        "fourteen",
        "fifteen",
        "sixteen",
        "seventeen",
        "eighteen",
        "nineteen"
      };
  private static final String[] toHundred =
      new String[] {
        null, null, "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"
      };

  String say(long number) {
    validate(number);
    if (number == 0L) return "zero";
    List<String> nouns = new LinkedList<>(Arrays.asList(scales));
    return recSay(number, nouns, new LinkedList<String>());
  }

  private static String recSay(long number, List<String> nouns, List<String> result) {
    if (number == 0) {
      return String.join(" ", result);
    }
    String scale = nouns.removeFirst();
    String abc = translate((int) (number % 1000));
    if (abc != null) {
      result.addFirst(join(abc, scale, " "));
    }
    return recSay(number / 1000, nouns, result);
  }

  private static String translate(int number) {
    String hundredString = getHundred(number);
    String tenString = getTen(number);
    return join(hundredString, tenString, " ");
  }

  private static String join(String first, String second, String delimiter) {
    if (first == null) {
      if (second == null) return null;
      return second;
    }
    if (second == null) {
      return first;
    }
    return first + delimiter + second;
  }

  private static String getHundred(int number) {
    if (number < 100) return null;
    String countHundred = toTwenty[number / 100];
    return String.format("%s hundred", countHundred);
  }

  private static String getTen(int number) {
    number %= 100;
    if (number < 20) {
      return toTwenty[number];
    }
    String countTen = toHundred[number / 10];
    String countOne = toTwenty[number % 10];
    return join(countTen, countOne, "-");
  }

  private static void validate(long number) {
    if (number < 0 || number > 999_999_999_999L) {
      throw new IllegalArgumentException("number must be in range [0, 1_000_000_000_000)");
    }
  }
}

File: python/say/say.py
import names

ZERO = 0
TEN = 10
TWENTY = TEN + TEN
HUNDRED = TEN * TEN
THOUSAND = TEN * HUNDRED
LIMIT = THOUSAND * THOUSAND * THOUSAND * THOUSAND


def say(number: int) -> str:
    if number < ZERO or number >= LIMIT or not isinstance(number, int):
        raise ValueError(names.OUT_OF_RANGE)
    if number == ZERO:
        return names.ZERO
    return names.SEPARATOR.join(chunks_of_thousands(number))


def chunks_of_thousands(number: int) -> list[str]:
    assert ZERO < number < LIMIT
    result = []
    for tillion in names.TILLIONS:
        mode_thousand = number % THOUSAND
        if mode_thousand:
            hundreds = chunk_hundreds(mode_thousand)
            if tillion:
                hundreds.append(tillion)
            result = hundreds + result  # reverse order
        number //= THOUSAND
    return result


def chunk_hundreds(mode_thousand: int) -> list[str]:
    assert ZERO < mode_thousand < THOUSAND
    result = []
    if mode_thousand >= HUNDRED:
        result.append(names.TENS[mode_thousand // HUNDRED])
        result.append(names.HUNDRED)
    mode_hundred = mode_thousand % HUNDRED
    if mode_hundred:
        result.append(chunk_tens(mode_hundred))
    return result


def chunk_tens(mode_hundred: int) -> str:
    assert ZERO < mode_hundred < HUNDRED
    if mode_hundred < TEN:
        return names.TENS[mode_hundred]
    if mode_hundred < TWENTY:
        return names.TEEN[mode_hundred]
    if mode_hundred % TEN == ZERO:
        return names.TIES[mode_hundred // TEN]
    return f"{names.TIES[mode_hundred//TEN]}-{names.TENS[mode_hundred%TEN]}"
